service/discovery/aws/storage.go

```diff
@@ -0,0 +1,382 @@
+/*
+ * Copyright 2021 Fraunhofer AISEC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *           $$\                           $$\ $$\   $$\
+ *           $$ |                          $$ |\__|  $$ |
+ *  $$$$$$$\ $$ | $$$$$$\  $$\   $$\  $$$$$$$ |$$\ $$$$$$\    $$$$$$\   $$$$$$\
+ * $$  _____|$$ |$$  __$$\ $$ |  $$ |$$  __$$ |$$ |\_$$  _|  $$  __$$\ $$  __$$\
+ * $$ /      $$ |$$ /  $$ |$$ |  $$ |$$ /  $$ |$$ |  $$ |    $$ /  $$ |$$ | \__|
+ * $$ |      $$ |$$ |  $$ |$$ |  $$ |$$ |  $$ |$$ |  $$ |$$\ $$ |  $$ |$$ |
+ * \$$$$$$\  $$ |\$$$$$   |\$$$$$   |\$$$$$$  |$$ |  \$$$   |\$$$$$   |$$ |
+ *  \_______|\__| \______/  \______/  \_______|\__|   \____/  \______/ \__|
+ *
+ * This file is part of Clouditor Community Edition.
+ */
+
+package aws
+
+import (
+	"clouditor.io/clouditor/api/discovery"
+	"clouditor.io/clouditor/voc"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+	"github.com/aws/smithy-go"
+	"time"
+)
+
+// awsS3Discovery handles the AWS API requests regarding the S3 service
+// ToDo: Generalize from s3 to storage for including other types, like EFS -> Probably storage folder with 3 .go files
+type awsS3Discovery struct {
+	client        S3API
+	isDiscovering bool
+	buckets       []bucket
```

Comment:
is there a reason to store the results here? the discovery should not do that in my opinion

Reply:
There is one API call that returns all buckets, or better say, bucket names. And other API calls, like getRegion(bucket string), require the name of the bucket to retrieve the appropriate values for the response. The idea is (as you can see in the List() method): Get all bucket names with getBuckets, store them in the struct awsS3Discovery. And the other API methods having this struct as receiver can perform their operations by accessing the names in the struct receiver.

But as I think about: Since only List() is exported and relevant for the discoverer, we may could also store the names in a local variable in List().